helix-term/src/commands/typed.rs

```diff
@@ -18,7 +18,36 @@ pub struct TypableCommand {
     pub doc: &'static str,
     // params, flags, helper, completer
     pub fun: fn(&mut compositor::Context, &[Cow<str>], PromptEvent) -> anyhow::Result<()>,
-    pub completer: Option<Completer>,
+    /// What completion methods, if any, does this command have?
+    pub signature: CommandSignature,
+}
+
+impl TypableCommand {
+    fn completer_for_argument_number(&self, n: usize) -> Option<&Completer> {
+        self.signature
+            .positional_args
+            .get(n)
+            .unwrap_or(&self.signature.var_args)
+            .as_ref()
+    }
+}
+
+#[derive(Clone)]
+pub struct CommandSignature {
+    // Arguments with specific completion methods based on their position.
+    positional_args: &'static [Option<Completer>],
+
+    // All remaining arguments will use this completion method, if set.
+    var_args: Option<Completer>,
```

Comment:
Actually thinking about it more I don't think we need to use `Option<Completer>` here we can just use `ui::completers::none` instead.
This will allow creating some utility functions (see below) to make the API a bit cleaner.